src/library/scala/collection/immutable/List.scala

```diff
@@ -394,6 +456,24 @@ sealed abstract class List[+A]
 
   final override def toList: List[A] = this
 
+  // Override for performance
+  override def equals(o: scala.Any): Boolean = {
+    @tailrec def listEq(a: List[_], b: List[_]): Boolean =
+      (a eq b) || {
+        if (a.nonEmpty && b.nonEmpty && a.head == b.head) {
+          listEq(a.tail, b.tail)
+        }
+        else {
+          a.isEmpty && b.isEmpty
+        }
+      }
+
```

Comment:
What's with all the superfluous braces?  Also, the logic can be streamlined a bit:
```
if (a.isEmpty) b.isEmpty
else if (b.nonEmpty && a.head == b.head) listEq(a.tail, b.tail)
else false
```

Reply:
I borrowed the code from https://github.com/scala/scala/blob/878e7d3e0d14633d19bac47dc9b532a54eab6379/src/library/scala/collection/LinearSeq.scala#L131-L146